src/ripple/legacy/0.27/CreateOffer.cpp

```diff
@@ -0,0 +1,631 @@
+        bool const canonical_gt (mTxnAccountID > issue.account);
+
+        bool const is_authorized (trustLine->getFieldU32 (sfFlags) &
+            (canonical_gt ? lsfLowAuth : lsfHighAuth));
+
+        if (!is_authorized)
+        {
+            if (m_journal.debug) m_journal.debug <<
+                "delay: can't receive IOUs from issuer without auth.";
+
+            return (mParams & tapRETRY)
+                ? terNO_AUTH
+                : tecNO_AUTH;
+        }
+    }
+
+    return tesSUCCESS;
+}
+
+std::pair<TER, legacy::core::Amounts>
+CreateOffer::crossOffers (legacy::core::LedgerView& view,
+    legacy::core::Amounts const& taker_amount)
+{
+    legacy::core::Taker::Options const options (mTxn.getFlags());
+
+    legacy::core::Clock::time_point const when (
+        mEngine->getLedger ()->getParentCloseTimeNC ());
+
+    legacy::core::LedgerView view_cancel (view.duplicate());
+    legacy::core::OfferStream offers (
+        view, view_cancel,
+        Book (taker_amount.in.issue(), taker_amount.out.issue()),
+        when, m_journal);
+    legacy::core::Taker taker (offers.view(), mTxnAccountID, taker_amount, options);
+
+    TER cross_result (tesSUCCESS);
+
+    while (true)
+    {
+        // Modifying the order or logic of these
+        // operations causes a protocol breaking change.
+
+        // Checks which remove offers are performed early so we
+        // can reduce the size of the order book as much as possible
+        // before terminating the loop.
+
+        if (taker.done())
+        {
+            m_journal.debug << "The taker reports he's done during crossing!";
+            break;
+        }
+
+        if (! offers.step ())
+        {
+            // Place the order since there are no
+            // more offers and the order has a balance.
+            m_journal.debug << "No more offers to consider during crossing!";
+            break;
+        }
+
+        auto const& offer (offers.tip());
+
+        if (taker.reject (offer.quality()))
+        {
+            // Place the order since there are no more offers
+            // at the desired quality, and the order has a balance.
+            break;
+        }
+
+        if (offer.account() == taker.account())
+        {
+            // Skip offer from self. The offer will be considered expired and
+            // will get deleted.
+            continue;
+        }
+
+        if (m_journal.debug) m_journal.debug <<
+                "  Offer: " << offer.entry()->getIndex() << std::endl <<
+                "         " << offer.amount().in << " : " << offer.amount().out;
+
+        cross_result = taker.cross (offer);
+
+        if (cross_result != tesSUCCESS)
+        {
+            cross_result = tecFAILED_PROCESSING;
+            break;
+        }
+    }
+
+    return std::make_pair(cross_result, taker.remaining_offer ());
+}
+
+CreateOffer::CreateOffer (STTx const& txn,
+    TransactionEngineParams params, TransactionEngine* engine)
```

Comment:
Just out of curiosity - why the raw pointer?  is it nullable?  What does it mean when the engine is null?